### PR TITLE
Allow form upload via admin UI

### PR DESCRIPTION
**Pull request by alxndrsn**
Allow form upload via admin UI

Add a basic UI for uploading forms via the admin's Configuration screen.  UI follows the pattern set on the same page for JSON form uploads.

This form has some specific limitations:

* form internal ID and couch doc ID must correspond, specifically:
    `doc._id = 'form:' + doc.internalId`
* form IDs cannot contain upper case letters
* form context.json file must be supplied
* icons cannot be uploaded along side the form

Issue: #3433

---

# Translation

On merge, the following translation keys should be renamed:

* `Upload Forms` -> `upload.json.forms.title`
* `Upload forms help` -> `upload.json.forms.help`

### CODE REVIEW

File: static/js/controllers/configuration-forms.js [hunk 1]
```diff
@@ -28,24 +30,102 @@ var moment = require('moment'),
...
+      var uploadXForm = function() {
+        $scope.xform = {
+          uploading: true,
+          error: false,
+          success: false,
+        };
+
+        var formFiles = $('#forms-upload-xform .form.uploader')[0].files;
+        if (!formFiles || formFiles.length === 0) {
+          uploadXFormFinished(new Error('XML file not found'));
```
Comment (garethbowen) [inline static/js/controllers/configuration-forms.js]: Return here so it doesn't try to access formFiles[0] and throw an exception (line 73).

File: static/js/controllers/configuration-forms.js [hunk 2]
```diff
@@ -28,24 +30,102 @@ var moment = require('moment'),
...
+        }
+
+        var contextFiles = $('#forms-upload-xform .context.uploader')[0].files;
+        if (!contextFiles || contextFiles.length === 0) {
+          uploadXFormFinished(new Error('JSON context file not found'));
```
Comment (garethbowen) [inline static/js/controllers/configuration-forms.js]: Same as above.

File: static/js/controllers/configuration-forms.js [hunk 3]
```diff
@@ -28,24 +30,102 @@ var moment = require('moment'),
...
+            var $xml = $($.parseXML(xml));
+            var title = $xml.find('title').text();
+            var formId = $xml.find('instance').children().first().attr('id');
+
+            var update = function(doc) {
+              doc.context = context;
+              doc.title = title;
+              doc.type = 'form';
+              doc.internalId = formId;
+              if (!doc._attachments) {
```
Comment (garethbowen) [inline static/js/controllers/configuration-forms.js]: Use the new `AddAttachment` service to clean this up slightly.

File: static/js/controllers/configuration-forms.js [hunk 4]
```diff
@@ -28,24 +30,102 @@ var moment = require('moment'),
...
+              }
+              doc._attachments.xml = {
+                content_type: 'application/xml',
+                data: new Blob([xml]),
+              };
+              return doc;
+            };
+
+            var couchId = 'form:' + formId;
+            DB().get(couchId, { include_attachments:true })
```
Comment (garethbowen) [inline static/js/controllers/configuration-forms.js]: This block could be cleaned up a lot. Try this...

```js
DB().get(couchId, { include_attachments:true })
  .catch(function(err) {
    if (err.status === 404) {
      return { _id: couchId };
    }
    throw err;
  })
  .then(function(doc) {
    return DB().put(update(doc));
  })
  .then(function() {
    uploadXFormFinished();
  })
  .catch(uploadXFormFinished);
```

I think this is functionally equivalent and I find it much easier to read (less nesting, fewer exit points).

File: translations/messages-en.properties [hunk 5]
```diff
@@ -286,12 +286,14 @@ Replace\ country\ code\ for = Replace country code for
 Messaging\ window = Messaging window
 Send\ scheduled\ messages\ between = Send scheduled messages between
 and = and
-Upload\ Forms = Upload forms
```
Comment (garethbowen) [inline translations/messages-en.properties]: Deleting translation keys here doesn't work as you'd expect (they'll keep coming back). Read the [translation documentation](https://github.com/medic/medic-docs/blob/master/development/translations.md#modifying-or-removing-translation-keys).

File: translations/messages-en.properties [hunk 6]
```diff
@@ -286,12 +286,14 @@ Replace\ country\ code\ for = Replace country code for
...
+Upload\ JSON\ Forms = Upload JSON forms
+Upload\ XForm = Upload XML form
 Upload\ failed = Upload failed
 Upload\ succeeded = Upload succeeded
 Installed\ Forms = Installed forms
 Choose\ file = Choose file
-Upload\ forms\ help = The file should contain a JSON encoded list of form definitions. Any existing forms will be overwritten.
```
Comment (garethbowen) [inline translations/messages-en.properties]: As above.

File: templates/partials/configuration_forms.html [hunk 7]
```diff
@@ -1,18 +1,45 @@
 <div class="col-sm-12">
-  <form id="forms-upload-form">
+  <form id="forms-upload-xform">
     <div>
       <fieldset>
-        <legend translate>Upload Forms</legend>
-        <div class="alert alert-danger alert-dismissible" role="alert" ng-show="error">
+        <legend translate>Upload XForm</legend>
```
Comment (garethbowen) [inline templates/partials/configuration_forms.html]: If you're going to change the translation key, switch to the new format. Maybe `xform.upload`?

File: templates/partials/configuration_forms.html [hunk 8]
```diff
@@ -1,18 +1,45 @@
...
+        <div class="alert alert-danger alert-dismissible" role="alert" ng-show="xform.error">
           <button type="button" class="close" data-dismiss="alert" aria-label="Close"><span aria-hidden="true">&times;</span></button>
           <span translate>Upload failed</span>
         </div>
-        <div class="alert alert-success" role="alert" ng-show="success" translate>Upload succeeded</div>
-        <button type="button" class="btn btn-default choose" ng-disabled="uploading">
+        <div class="alert alert-success" role="alert" ng-show="xform.success" translate>Upload succeeded</div>
+        <label>XML
```
Comment (garethbowen) [inline templates/partials/configuration_forms.html]: Missing internationalisation.

File: templates/partials/configuration_forms.html [hunk 9]
```diff
@@ -1,18 +1,45 @@
...
+          <input class="form uploader" type="file" accept="application/xml"/>
+        </label>
+        <label> Context
```
Comment (garethbowen) [inline templates/partials/configuration_forms.html]: Missing internationalisation.

File: templates/partials/configuration_forms.html (hunk 8 above)
Comment (alxndrsn) [inline templates/partials/configuration_forms.html]: What's the French for XML?

File: static/js/controllers/configuration-forms.js (hunk 3 above)
Comment (alxndrsn) [inline static/js/controllers/configuration-forms.js]: nice

File: static/js/controllers/configuration-forms.js (hunk 4 above)
Comment (alxndrsn) [inline static/js/controllers/configuration-forms.js]: :+1: :+1: :+1:

Comment (alxndrsn) [inline static/js/controllers/configuration-forms.js]: Done, but this comment is on the one line that didn't change :¬)

**Comment by alxndrsn** [comment]
@garethbowen ready for another review

**Comment by garethbowen** [comment]
Nice one. It appears as though there are still a couple of unanswered comments:

https://github.com/medic/medic-webapp/pull/3451/files#r115154933
https://github.com/medic/medic-webapp/pull/3451/files#r115154935

Other than that, all good!

File: static/js/controllers/configuration-forms.js (hunk 1 above)
Comment (alxndrsn) [inline static/js/controllers/configuration-forms.js]: This has been dealt with without modifying this line.  See subsequent line(s).

Comment (garethbowen) [inline static/js/controllers/configuration-forms.js]: Right you are!

**Comment by alxndrsn** [comment]
> Nice one. It appears as though there are still a couple of unanswered comments:
>
> https://github.com/medic/medic-webapp/pull/3451/files#r115154933
> https://github.com/medic/medic-webapp/pull/3451/files#r115154935

Both addressed already in a way that github review doesn't notice :-\

**Pull request merged**